File: Next Greater Element II/next greater element II.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        stack<int> stack;
        vector<int> ans(nums.size(), -1);
        stack.push(0);
        
        for(int i = 1; i < nums.size(); i++) {
            int current = nums[i];
            while (stack.size() > 0 && nums[stack.top()] < current) {
                ans[stack.top()] = current;
                stack.pop();
            }
            stack.push(i);
        }
        
        for(int i = 0; i < nums.size(); i++) {
            int current = nums[i];
            while (stack.size() > 0 && nums[stack.top()] < current) {
                ans[stack.top()] = current;
                stack.pop();
            }
        }
        
        return ans;
    }
};

File: Next Greater Element II/next greater element II.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        stack = []
        stack.append(0)
        n = len(nums)
        ans = [-1]*n
        
        
        for i in range(1, n):
            current = nums[i]
            while (len(stack) > 0 and nums[stack[-1]] < current):
                ans[stack[-1]] = current
                stack.pop()
            stack.append(i)
        
        
        for i in range(0, n):
            current = nums[i]
            while (len(stack) > 0 and nums[stack[-1]] < current):
                ans[stack[-1]] = current
                stack.pop()
        
        return ans